File: leoz-node/src/main/java/org/deku/leoz/node/data/sync/EntityPublisher.java
package org.deku.leoz.node.data.sync;

import com.google.common.base.Stopwatch;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leoz.config.messaging.ActiveMQConfiguration;
import org.deku.leoz.config.messaging.MessagingConfiguration;
import org.deku.leoz.node.data.repositories.EntityRepository;
import org.deku.leoz.node.data.sync.v1.EntityStateMessage;
import org.deku.leoz.node.data.sync.v1.EntityUpdateMessage;
import org.eclipse.persistence.queries.ScrollableCursor;
import sx.jms.Channel;
import sx.jms.converters.DefaultConverter;
import sx.jms.listeners.SpringJmsListener;

import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.MessageProducer;
import javax.jms.Session;
import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import java.sql.Timestamp;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

/**
 * Created by masc on 18.06.15.
 */
public class EntityPublisher extends SpringJmsListener {
    private Log mLog = LogFactory.getLog(this.getClass());
    /** Messaging context */
    private MessagingConfiguration mMessagingConfiguration;
    /** Entity manager factory */
    private EntityManagerFactory mEntityManagerFactory;

    /**
     * c'tor
     * @param messagingConfiguration
     * @param entityManagerFactory
     */
    public EntityPublisher(MessagingConfiguration messagingConfiguration, EntityManagerFactory entityManagerFactory) {
        super(
                messagingConfiguration.getBroker().getConnectionFactory(),
                messagingConfiguration::getCentralEntitySyncQueue,
                new DefaultConverter(
                        DefaultConverter.SerializationType.KRYO,
                        DefaultConverter.CompressionType.GZIP));
        mMessagingConfiguration = messagingConfiguration;
        mEntityManagerFactory = entityManagerFactory;
    }

    /**
     * Publish entity update notification
     * @param entityType
     * @param timestamp
     */
    public void publish(Class entityType, Timestamp timestamp) throws JMSException {
        try (Channel mc = new Channel(
                ActiveMQConfiguration.getInstance().getBroker().getConnectionFactory(),
                ActiveMQConfiguration.getInstance().getNodeEntitySyncTopic(),
                this.getConverter(),
                false,
                Channel.DeliveryMode.NonPersistent,
                Duration.ofMinutes(5),
                null,
                Duration.ofSeconds(10))) {

            mc.send(new EntityStateMessage(entityType, timestamp));
        }
    }

    @Override
    public void onMessage(Message message, Session session) throws JMSException {
        mLog.debug(String.format("Message id [%s] %s",
                message.getJMSMessageID(),
                LocalDateTime.ofInstant(
                        Instant.ofEpochMilli(
                                message.getJMSTimestamp()), ZoneId.systemDefault())));

        Stopwatch sw = Stopwatch.createStarted();

        // Create new message converter for this session, just for clean statistics sake
        DefaultConverter messageConverter = (DefaultConverter)this.getConverter();

        // Entity state message
        EntityStateMessage esMessage = (EntityStateMessage) messageConverter.fromMessage(message);
        Class entityType = esMessage.getEntityType();
        Timestamp timestamp = esMessage.getTimestamp();
        Function<String, String> lfmt = s -> "[" + entityType.getCanonicalName() + "]" + " " + s;


        EntityManager em = mEntityManagerFactory.createEntityManager();
        EntityRepository er = new EntityRepository(em, entityType);

        // Count records
        Long count = er.countNewerThan(timestamp);

        EntityUpdateMessage euMessage = new EntityUpdateMessage(count);
        mLog.debug(euMessage);

        MessageProducer mp = session.createProducer(message.getJMSReplyTo());
        mp.send(messageConverter.toMessage(euMessage, session));

        if (count > 0) {
            // Query with cursor
            ScrollableCursor cursor = er.findNewerThan(timestamp);

            final int CHUNK_SIZE = 500;
            ArrayList buffer = new ArrayList(CHUNK_SIZE);
            mLog.info(lfmt.apply(String.format("Sending %d", count)));
            while (true) {
                Object next = null;
                if (cursor.hasNext()) {
                    next = cursor.next();
                    buffer.add(next);
                }
                if (buffer.size() >= CHUNK_SIZE || next == null) {
                    if (buffer.size() > 0) {
                        //mLog.debug(String.format("Sending %d", buffer.size()));
                        mp.send(messageConverter.toMessage(buffer.toArray(),session));
                        buffer.clear();
                    }

                    if (next == null)
                        break;
                }
            }

            // Send empty array -> EOS
            Message eosMsg = messageConverter.toMessage(new Object[0], session);
            eosMsg.setBooleanProperty(EntityUpdateMessage.EOS_PROPERTY, true);
                    mp.send(eosMsg);
        }
        session.commit();
        mLog.info(lfmt.apply(String.format("Sent %d in %s (%d bytes)", count, sw, messageConverter.getBytesWritten())));

        em.close();
    }
}

File: leoz-node/src/main/java/org/deku/leoz/node/data/sync/EntityPublisher.kt
package org.deku.leoz.node.data.sync

import com.google.common.base.Stopwatch
import org.deku.leoz.config.messaging.ActiveMQConfiguration
import org.deku.leoz.node.data.repositories.EntityRepository
import org.deku.leoz.node.messaging.entities.EntityStateMessage
import org.deku.leoz.node.messaging.entities.EntityUpdateMessage
import org.eclipse.persistence.queries.ScrollableCursor
import sx.jms.Channel
import sx.jms.Handler
import sx.jms.listeners.SpringJmsListener
import java.sql.Timestamp
import java.util.*
import java.util.concurrent.Executor
import javax.jms.JMSException
import javax.persistence.EntityManager
import javax.persistence.EntityManagerFactory

/**
 * Entity publisher
 * Created by masc on 18.06.15.
 * @param messagingConfiguration
 * @param entityManagerFactory
 */
class EntityPublisher(
        private val notificationChannelConfiguration: Channel.Configuration,
        /** Entity manager factory  */
        private val entityManagerFactory: EntityManagerFactory,
        executor: Executor)
:
        SpringJmsListener({ Channel(notificationChannelConfiguration) }, executor),
        Handler<EntityStateMessage> {
    init {
        this.addDelegate(this)
    }

    /**
     * Publish entity update notification
     * @param entityType
     * *
     * @param timestamp
     */
    @Throws(JMSException::class)
    fun publish(entityType: Class<*>, syncId: Long?) {
        Channel(ActiveMQConfiguration.instance.entitySyncTopic).use {
            val msg = EntityStateMessage(entityType, syncId)
            log.info("Publishing [${msg}]")
            it.send(msg)
        }
    }

    @Throws(JMSException::class)
    override fun onMessage(message: EntityStateMessage, replyChannel: Channel?) {
        var em: EntityManager? = null
        try {
            val sw = Stopwatch.createStarted()

            // Entity state message
            val esMessage = message
            val entityType = esMessage.entityType
            val syncId = esMessage.syncId
            val lfmt = { s: String -> "[" + entityType!!.canonicalName + "]" + " " + s }

            em = entityManagerFactory.createEntityManager()
            val er = EntityRepository(em, entityType)

            // Count records
            val count = er.countNewerThan(syncId)

            replyChannel!!.statistics.enabled = true

            // Send entity update message
            val euMessage = EntityUpdateMessage(count)
            log.debug(lfmt(euMessage.toString()))
            replyChannel.send(euMessage)

            if (count > 0) {
                // Query with cursor
                var cursor: ScrollableCursor? = null
                try {
                    cursor = er.findNewerThan(syncId)

                    val CHUNK_SIZE = 500
                    val buffer = ArrayList<Any?>(CHUNK_SIZE)
                    log.info(lfmt("Sending ${count}"))
                    while (true) {
                        var next: Any? = null
                        if (cursor.hasNext()) {
                            next = cursor.next()
                            buffer.add(next)
                        }
                        if (buffer.size >= CHUNK_SIZE || next == null) {
                            if (buffer.size > 0) {
                                replyChannel.send(buffer.toArray())
                                buffer.clear()
                            }

                            if (next == null)
                                break
                        }
                    }

                    // Send empty array -> EOS
                    replyChannel.send(arrayOfNulls<Any>(0), messageConfigurer = {
                        it.setBooleanProperty(EntityUpdateMessage.EOS_PROPERTY, true)
                    })
                } finally {
                    if (cursor != null)
                        cursor.close()
                }
            }
            log.info(lfmt("Sent ${count} in ${sw} (${replyChannel.statistics.bytesSent} bytes)"))
        } catch(e: Exception) {
            log.error(e.message, e);
        } finally {
            if (em != null)
                em.close()
        }
    }
}
